### PR TITLE
Abc

### PR DESCRIPTION
**Pull request by cdonovick**
Adds `bit_vector_abc.py` with classes `AbstractBitVector` and `AbstractBitVectorMeta`

Forces BitVector types to be bound to a specific size using the syntax `BitVectorT[size]`

Construction of BitVectors from `int` with implicit size  is still supported

Uses "abstractness" of the method `_size` to determine if a BitVector type is bound to a size.  It might be more appropriate to add some other attribute or inherit from some marker class instead.

### CODE REVIEW

**Comment by cdonovick** [comment]
@leonardt @phanrahan

**Comment by cdonovick** [comment]
Should be noted `AbstractBitVectorMeta` is really `AbstractFixedSizedMeta` and really doesn't do anything but construct classes with the a method called `_size`

File: bit_vector/bit_vector.py [hunk 1]
```diff
@@ -159,64 +158,55 @@ def __setitem__(self, index, value):
     @no_x
     def __getitem__(self, index):
         if isinstance(index, slice):
-            return BitVector(self._bits[index])
+            v = self._bits[index]
+            return BitVector[len(v)](v)
         else:
-            return self.bits()[index]
+            return BitVector[1](self.bits()[index])
```
Comment (leonardt) [inline bit_vector/bit_vector.py]: Should this be a `Bit` object instead? (We will have to define this)

File: bit_vector/bit_vector.py [hunk 2]
```diff
@@ -159,64 +158,55 @@ def __setitem__(self, index, value):
...
 
     @binary
     def bvlshr(self, other):
-        return BitVector( self.as_uint() >> other.as_uint(), num_bits=self.num_bits)
+        return type(self)( self.as_uint() >> other.as_uint())
 
     @binary
     def bvashr(self, other):
-        return BitVector( self.as_sint() >> other.as_uint(), num_bits=self.num_bits)
+        return type(self)( self.as_sint() >> other.as_uint())
```
Comment (leonardt) [inline bit_vector/bit_vector.py]: We could clean this up by moving all casting logic to a decorator (or add it to the existing decorator), but I don't think this is a blocking issue.

File: bit_vector/bit_vector.py [hunk 3]
```diff
@@ -239,54 +229,24 @@ def bvcomp(self, other):
...
-        return BitVector(self.as_sint() <= other.as_sint(), num_bits=1)
-
-    @binary_no_cast
-    def bvsgt(self, other):
-        return BitVector(self.as_sint() > other.as_sint(), num_bits=1)
-
-    @binary
-    def bvsge(self, other):
-        return BitVector(self.as_sint() >= other.as_sint(), num_bits=1)
+        return BitVector[1](self.as_sint() < other.as_sint())
```
Comment (leonardt) [inline bit_vector/bit_vector.py]: Did we mean to drop all these comparisons?

Comment (leonardt) [reply bit_vector/bit_vector.py]: Nevermind, these are defined in the abstract class in terms of other methods which have been implemented

File: bit_vector/bit_vector.py [hunk 4]
```diff
@@ -296,84 +256,74 @@ def adc(a, b, c):
...
-        return BitVector(self.as_uint() * other.as_uint(), num_bits=self.num_bits)
+        return type(self)(self.as_uint() * other.as_uint())
 
     @binary
     def bvudiv(self, other):
         other = other.as_uint()
         if other == 0:
-            return BitVector(1, num_bits=self.num_bits)
-        return BitVector(self.as_uint() // other, num_bits=self.num_bits)
+            return type(self)((1 << self.size) - 1)
```
Comment (leonardt) [inline bit_vector/bit_vector.py:278]: Is this the expected semantics for div by zero? I remember you mentioned it was a special case, just wanted to double check.

File: bit_vector/bit_vector.py [hunk 5]
```diff
@@ -440,7 +387,8 @@ def sext(self, other):
 
         TODO: Do this implicit conversion for built-in types like UIntVector.
         """
-        return self.concat(BitVector(other.as_uint() * [self[-1]]), self)
+        ext = other.as_uint()
+        return self.concat(BitVector[ext](ext * [self[-1]]), self)
```
Comment (leonardt) [inline bit_vector/bit_vector.py]: I think this is changing the interface to `sext`. Before, `other` indicated how many bits to extend the current bit vector. Now it uses `ext` as the size of the target bit vector. We should decide on a semantics since if we want to change it to be this, this is a good time to do it (also is this the SMT semantics?).

**Review by leonardt — COMMENTED** [review]
Some comments, still need to read through the abstract class definition

File: bit_vector/bit_vector.py (hunk 1 above)
Comment (cdonovick) [reply bit_vector/bit_vector.py]: I will add a bit type

File: bit_vector/bit_vector.py (hunk 4 above)
Comment (cdonovick) [reply bit_vector/bit_vector.py:278]: This is the SMT-lib semantics

File: bit_vector/bit_vector.py (hunk 5 above)
Comment (cdonovick) [reply bit_vector/bit_vector.py]: The SMT-lib zero and sign extend have the semantics we currently have.

File: bit_vector/bit_vector.py (hunk 2 above)
Comment (cdonovick) [reply bit_vector/bit_vector.py]: I dislike this because it makes the behavior kind of opaque in my mind.

**Review by cdonovick — COMMENTED** [review]
I am not in love with how I am currently supporting the type family concept but it gets the job done.

File: bit_vector/bit_vector_abc.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+    def __getitem__(cls, idx : int) -> 'AbstractBitVectorMeta':
+        try:
+            return AbstractBitVector._class_cache[cls, idx]
+        except KeyError:
+            pass
+
+        if not isinstance(idx, int):
+            raise TypeError()
+        if idx < 0:
+            raise ValueError('Size of BitVectors must be positive')
```
Comment (leonardt) [inline bit_vector/bit_vector_abc.py:71]: Did we decide to allow size 0?

**Review by leonardt — APPROVED** [review]
LGTM, finally had a chance to read the ABC implementation, makes enough sense that I think could help debug any issues that come up. Let ship this!

File: bit_vector/bit_vector_abc.py (hunk 6 above)
Comment (cdonovick) [reply bit_vector/bit_vector_abc.py:71]: I think we decided against it but forbidding them broke some tests and create weird corner cases for user code.  I think its better to allow them.

Comment (leonardt) [reply bit_vector/bit_vector_abc.py:71]: Okay, let's leave it for now but keep on eye on whether this becomes a problem.
